### ISSUE TITLE
can't login with Azure AD using Microsoft Edge browser

### ISSUE THREAD

**Issue by carlin-q-scott** [Azure/vscode-aks-tools#210]
**Describe the bug**
I can't log in using the user flow because MS Edge Browser blocks localhost on random ports. kubectl on the command line correctly uses device code login.

**To Reproduce**
1. Set MS Edge as default browser
2. Create AKS cluster with Azure AD auth
3. Log in to cluster using this extension
4. Get stuck in a login error loop

**Expected behavior**
Device code login flow dialog works.

**Actual behavior**
Extension attempts to log in using user flow several times, and then falls back to device code flow. But the user flow dialog continues, and I guess cancelling it or closing the VS Code instance results in it clearing the auth token granted through the device flow. I'm not sure exactly what happens, but the extension continues acting like I'm not logged in, even though I completed the device code flow successfully. Attempting to perform actions using the Kubernetes extension results in 4 browser tabs opening the user flow again.

**Workaround**

1. Open terminal
2. `kubectl config use-context cluster-name`
3. `kubectl get services`
4. Follow device code login instructions presented by kubelogin
5. I'm now able to use the extension

**Desktop (please complete the following information):**
 - OS: Windows 11
 - Browser: Edge
 - Version: 112.0.1722.68

**Comment by Tatsinnit** [comment]
Thank you so much @carlin-q-scott, we are looking into this and will get back you soon. Really appreciate this. ❤️☕️ cc: @peterbom

**Comment by peterbom** [comment]
@carlin-q-scott: thanks for the thorough description. I think the key thing here is Edge blocking the redirect to localhost. That's something I'm not experiencing here, which would explain why I've never noticed this bug. It also means I haven't yet figured out how to reproduce it, so I'd really appreciate your help with the following...

I'd like to understand better what's happening with Edge here. I see it [disables 'localhost loopback'](https://stackoverflow.com/questions/40578564/microsoft-edge-blocks-local-server) by default. But I thought that'd only affect redirects to something like `127.0.0.1`, and `localhost` would be fine.

Can you share what you observe when the localhost address gets blocked? On my machine the browser redirects to `http://localhost:63227/?code=0.xxxxxxx` and I get a message saying "Authentication complete. You can return to the application. Feel free to close this browser tab." I assume this is the part that fails for you? Does that page simply fail to load? If you open up Developer Tools, do you see any console messages or failing network requests that might indicate how/why this part of the flow is failing?

As an aside, it is intentional that the extension is using the interactive flow rather than device code login. We designed it that way because for most cases this works best in a GUI environment like VS Code (we're invoking kubectl commands as background processes, so copy/pasting device codes doesn't really work because you never see them). I _think_ the reason your workaround is effective is that you're overriding the kubeconfig user from your terminal and performing the authentication from there. But when the token expires I think you'll experience the cluster explorer hanging as it waits for a device code login that never happens.

**Comment by carlin-q-scott** [comment]
I think my issue with the interactive login flow was a combination of a couple feature flags in Edge. 

MS Edge has removed the 'localhost loopback' feature flag mentioned in the SO answer you linked to, but they've added an "Allow invalid certificates for resources loaded from localhost" feature flag that was disabled for me. I believe enabling this will fix my issue, as the error I was being presented with was an invalid ssl error; I don't remember exactly which one. I have the HTTPS redirect feature flag enabled, so that's why it wasn't loading with http://localhost for me. I don't have this issue with my projects launched from VS because the dotnet cert authority is trusted by my development machine.

> we're invoking kubectl commands as background processes, so copy/pasting device codes doesn't really work because you never see them

I was presented with a dialog that automatically copied the device code to my clipboard and opened the auth website where I needed to paste the code. Is that coming from a different extension? It looked like the interactive flow dialog except the text content.

> But when the token expires I think you'll experience the cluster explorer hanging as it waits for a device code login that never happens.

That has not been my experience. I've only had to authorize twice in the past 12 months. And I just checked my current token, and it's only "good" for a little over an hour. It must be re-using the access token to get a new one.

**Comment by peterbom** [comment]
@carlin-q-scott - thank you for the extra info!

I'm trying to replicate your setup from edge://flags/ and I've verified that "Allow invalid certificates for resources loaded from localhost" is disabled, and also set "Automatic HTTPS" (which I was guessing to be the HTTPS redirect flag you mentioned?). My settings now look like this:
<img width="599" alt="image" src="https://github.com/Azure/vscode-aks-tools/assets/1817696/ef9db2a5-8514-405e-9593-6b992efa37e3">

I'm still not getting any localhost URLs redirected to HTTPS, though, so that sounds like the difference between our setups. They _should_ be just HTTP, so it'd make sense that you're getting certificate errors if it is trying to use HTTPS. Can you tell me any more about how you're making your browser redirect to HTTPS?

> I was presented with a dialog that automatically copied the device code to my clipboard and opened the auth website where I needed to paste the code. Is that coming from a different extension? It looked like the interactive flow dialog except the text content.

I'm not sure how that's happening! Can you check the user for your cluster in your `kubeconfig` file? Mine looks something like this (note the '--login interactive' in there). Is yours similar?
```yaml
users:
- name: my-aad-cluster
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1beta1
      args:
      - get-token
      - --environment
      - AzurePublicCloud
      - --server-id
      - ...
      - --client-id
      - ...
      - --tenant-id
      - ...
      - --login
      - interactive
      - --token-cache-dir
      - C:\Users\name\AppData\Local\Temp\tmp-...
      command: C:\Users\name\.vs-kubernetes\tools\kubelogin\v0.0.22\kubelogin.exe
      env: null
      interactiveMode: IfAvailable
      provideClusterInfo: false
```

**Comment by carlin-q-scott** [comment]
I found a resolution for my issue: [Microsoft Edge redirects http://localhost to https://localhost](https://stackoverflow.com/questions/63825407/microsoft-edge-redirects-http-localhost-to-https-localhost)

I fixed the issue by doing the following:
1. Remove localhost from my [HSTS cache](edge://net-internals/#hsts)
2. Remove AKS cluster from kube config
    I did this manually, but I could probably have done it through the kubernetes extension UI
3. Add the AKS cluster back to kube config
4. Everything works! I didn't even need to log in for some reason. And my kube config matches yours now.

I have HSTS and HTTPS redirect configured for the web apps I work on. So any one of them could have set the HTTPS redirect for localhost. I guess I will have to be careful to never navigate to them in Edge using HTTP while debugging locally.
